optuna/testing/storages.py

```diff
@@ -43,6 +45,49 @@
 SQLITE3_TIMEOUT = 300
 
 
+class FindFreePortLockFile:
+    if sys.platform == "win32":
+        _lock_path: str = os.path.join(
+            os.environ.get("PROGRAMDATA", "C:\\ProgramData"),
+            "optuna",
+            "optuna_find_free_port.lock",
+        )
+    else:
+        _lock_path: str = "/tmp/optuna_find_free_port.lock"
+
+    def __init__(self) -> None:
+        os.makedirs(os.path.dirname(self._lock_path), exist_ok=True)
+        self._lockfile = open(self._lock_path, "w")
+
+    def __enter__(self) -> "FindFreePortLockFile":
+        if sys.platform == "win32":
+            import msvcrt
+
+            msvcrt.locking(self._lockfile.fileno(), msvcrt.LK_LOCK, 1)
+        else:
+            import fcntl
+
+            fcntl.flock(self._lockfile, fcntl.LOCK_EX)
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException],
+        exc_val: BaseException,
+        exc_tb: TracebackType,
+    ) -> None:
+        if sys.platform == "win32":
+            import msvcrt
+
+            msvcrt.locking(self._lockfile.fileno(), msvcrt.LK_UNLCK, 1)
+        else:
+            import fcntl
+
+            fcntl.flock(self._lockfile, fcntl.LOCK_UN)
+
+        self._lockfile.close()
+
+
 class StorageSupplier:
```

Comment:
This is not a problem with your PR itself, but subclassing from `AbstractContextManager` might be better[^1]:
```suggestion
class StorageSupplier(AbstractContextManager):
```

Also, as @nabenabe0928 mentions, this class can be rewritten by using the `contextmanager` decorator. However, for this class, the current approach might be better, since the `__enter__` method is much more complicated than `FindFreePortLockFile`.

[^1]: https://docs.python.org/3/library/contextlib.html